fastparse/shared/src/test/scala/fastparse/iterator/IteratorTests.scala

```diff
@@ -0,0 +1,105 @@
+package fastparse.iterator
+
+import fastparse.{ElemTypeFormatter, IteratorParserInput}
+import fastparse.all._
+import utest._
+
+import scala.collection.mutable
+
+object IteratorTests extends TestSuite {
+
+  class LoggedDropsParserInput[ElemType](data: Iterator[IndexedSeq[ElemType]])
+                                        (implicit formatter: ElemTypeFormatter[ElemType])
+    extends IteratorParserInput[ElemType](data) {
+
+    var drops = mutable.Set.empty[Int]
+
+    override def dropBuffer(index: Int): Unit = {
+      drops.add(index)
+      super.dropBuffer(index)
+    }
+  }
+
+  def toInput(string: String) = new LoggedDropsParserInput[Char](string.grouped(1).map(wrapString))
+
+  val tests = TestSuite {
+    'cuts {
+      'capturing {
+        val p = P( "a" ~/ "b" ~/ "c")
+        val capt = P( p.! ~ p.! ~ p.!)
+        val input = toInput("abcabcabc")
+        val Parsed.Success(res, i) = capt.parseInput(input)
+        assert(
+          i == 9,
+          res == ("abc", "abc", "abc"),
+          Seq(3, 6, 9).forall(input.drops.contains)
+        )
+      }
+
+      'nocut {
+        val p = P( "a" ~/ "b" ~/ "c")
+        val nocut = P((NoCut(p) ~ NoCut(p) ~/ NoCut(p)) | "abcd")
+
+        val input1 = toInput("abcabcabc")
+        val Parsed.Success(_, i1) = nocut.parseInput(input1)
+        assert(
+          i1 == 9,
+          Seq(9).forall(input1.drops.contains)
+        )
+
+        val input2 = toInput("abcd")
+        val Parsed.Success(_, i2) = nocut.parseInput(input2)
+        assert(i2 == 4)
+      }
+
+      'either {
+        val p = P( "a" ~ "b" ~ "c")
+        val either = P( (p ~ End) | ("abc" ~ p ~ End) | ("abcabc" ~ p ~ End))
+
+        val input1 = toInput("abcabcabc")
+        val Parsed.Success(_, i1) = either.parseInput(input1)
+        assert(
+          i1 == 9,
+          Seq(6, 7, 8, 9).forall(input1.drops.contains)
```

Comment:
It's not obvious to me why there should be drops at indices 6, 7, 8 and 9 in this test. Presumably it's because it's the "last" item in the `|` chain, but a comment would do great here to explain that fact